proc/futex.c

```diff
@@ -0,0 +1,78 @@
+#include "hal/hal.h"
+#include "lib/lib.h"
+#include "include/errno.h"
+#include "process.h"
+#include "threads.h"
+#include "futex.h"
+
+/* TODO: Handle hash collisions */
+static futex_sleepqueue_t *futex_getSleepQueue(process_t *process, addr_t address)
+{
+	u32 key;
+	futex_sleepqueue_t *futex_sleepqueue;
+
+	proc_lockSet(&process->lock);
+	key = address >> 3;
+	key ^= key >> FUTEX_SLEEPQUEUES_BITS;
+	futex_sleepqueue = &process->futex_sleepqueues[key & FUTEX_SLEEPQUEUES_MASK];
+	proc_lockClear(&process->lock);
+	return futex_sleepqueue;
+}
+
+int proc_futexWait(u32 *address, u32 value, time_t timeout)
+{
+	process_t *current_process;
+	addr_t key;
+	spinlock_ctx_t ctx;
+	futex_sleepqueue_t *sleepqueue;
+	int err;
+	time_t now, total_timeout = 0;
+
+	if (*address != value) {
+		return -EAGAIN;
+	}
+	current_process = proc_current()->process;
+	key = (addr_t)address;
+	sleepqueue = futex_getSleepQueue(current_process, key);
+
+	if (timeout > 0) {
+		proc_gettime(&now, NULL);
+		total_timeout = now + timeout;
+	}
+
+	hal_spinlockSet(&sleepqueue->spinlock, &ctx);
+	sleepqueue->count += 1;
+	err = proc_threadWaitInterruptible(&sleepqueue->threads, &sleepqueue->spinlock, total_timeout, &ctx);
+	sleepqueue->count -= 1;
+	hal_spinlockClear(&sleepqueue->spinlock, &ctx);
+	return err;
+}
+
+int proc_futexWakeup(u32 *address, u32 n_threads)
+{
+	process_t *current_process;
+	addr_t key;
+	int i, woken = 0;
+	spinlock_ctx_t ctx;
+	futex_sleepqueue_t *sleepqueue;
+
+	if (n_threads == 0) {
+		return 0;
+	}
+
+	current_process = proc_current()->process;
+	key = (addr_t)address;
+	sleepqueue = futex_getSleepQueue(current_process, key);
+	hal_spinlockSet(&sleepqueue->spinlock, &ctx);
+	if (n_threads == FUTEX_WAKEUP_ALL) {
```

Comment:
Please check out `_proc_threadBroadcast` function (it wakes up all threads waiting for cond).